python/lsst/cp/pipe/cpLinearitySolve.py

```diff
@@ -1211,3 +1118,1241 @@ def debugFit(self, stepname, xVector, yVector, yModel, mask, ampName):
+
+    Parameters
+    ----------
+    ptc : `lsst.ip.isr.PhotonTransferCurveDataset`
+        Input PTC to do grouping.
+    doAutoGrouping : `bool`
+        Do automatic grouping of pairs?
+    autoGroupingUseExptime : `bool`
+        Use exposure time for automatic grouping of pairs?
+    autoGroupingMaxSignalFraction : `float`
+        All exposures with signal higher than this threshold will
+        be put into the largest signal group.
+    autoGroupingThreshold : `float`
+        Minimum relative jump from sorted values to determine a group.
+    minPhotodiodeCurrent : `float`
+        Minimum photodiode current if auto-grouping is used and
+        autoGroupingUseExptime is False.
+    splineGroupingColumn : `str` or `None`
+        Column to be used for spline grouping (if doAutoGrouping is False).
+
+    Returns
+    -------
+    groupingValues : `np.ndarray`
+        Array of values that are unique for a given group.
+    """
+    nPair = np.asarray(ptc.inputExpIdPairs[ptc.ampNames[0]]).shape[0]
+    groupingValues = np.zeros(nPair, dtype=np.int64)
+
+    if not doAutoGrouping:
+        if groupingColumn is not None:
+            if groupingColumn not in ptc.auxValues:
+                raise ValueError(f"Config requests grouping by {groupingColumn}, "
+                                 "but this column is not available in ptc.auxValues.")
+
+            uGroupValues = np.unique(ptc.auxValues[groupingColumn])
+            for i, uGroupValue in enumerate(uGroupValues):
+                groupingValues[ptc.auxValues[groupingColumn] == uGroupValue] = i
+    else:
+        means = np.zeros((nPair, len(ptc.ampNames)))
+        exptimes = np.zeros_like(means)
+        for i, ampName in enumerate(ptc.ampNames):
+            means[:, i] = ptc.rawMeans[ampName] * ptc.gain[ampName]
+            exptimes[:, i] = ptc.rawExpTimes[ampName]
+        detMeans = np.nanmean(means, axis=1)
+        detExptimes = np.nanmean(exptimes, axis=1)
+
+        if autoGroupingUseExptime:
+            abscissa = detExptimes
+        else:
+            abscissa = ptc.photoCharges[ptc.ampNames[0]].copy()
+            # Set illegal photocharges to NaN.
+            abscissa[abscissa < minPhotodiodeCurrent] = np.nan
+
+        ratio = detMeans / abscissa
+        ratio /= np.nanmedian(ratio)
+
+        # Adjust those that are above threshold so they fall into the
+        # largest group.
+        above = (detMeans > autoGroupingMaxSignalFraction*np.nanmax(detMeans))
+        maxIndex = np.argmax(detMeans[~above])
+        ratio[above] = ratio[maxIndex]
+
+        # The clustering of ratios into groups is performed with a simple
+        # algorithm based on sorting and looking for the largest gaps.
+        # See https://stackoverflow.com/a/18385795
+        st = np.argsort(ratio)
+        stratio = ratio[st]
+        delta = stratio[1:] - stratio[0: -1]
+
+        transitions, = np.where(delta > autoGroupingThreshold)
+        if len(transitions) > 0:
```

Comment:
Does this fail cleanly if no delta is above the threshold?  I guess everything is just then in group 0.